### ISSUE TITLE
Unable to create a nested folder structure more than 2 levels deep

### ISSUE THREAD

**Issue by cloud-dan** [terraform-providers/terraform-provider-azurerm#5398]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

* `Terraform v0.12.19`
* `+ provider.AzureRM v1.40.0`

### Affected Resource(s)

* `azurerm_storage_share_directory`

### Terraform Configuration Files

```hcl
locals {
  folder_sructure = [
    "Folder1",
    "Folder1/Folder2",
    "Folder1/Folder2/Folder3"
  ]
}

resource "azurerm_storage_share" "files" {
  name                 = "files"
  storage_account_name = data.azurerm_storage_account.Data.name
  quota                = 50
}

resource "azurerm_storage_share_directory" "folder_sructure" {
  for_each = toset(local.folder_sructure)

  name                 = each.key
  share_name           = azurerm_storage_share.files.name
  storage_account_name = data.azurerm_storage_account.data.name
}
```
### Panic Output

```
Error: name must contain only uppercase and lowercase alphanumeric characters, numbers and hyphens. It must start and end with a letter and end only with a number or letter

  on storage._shares.tf line 146, in resource "azurerm_storage_share_directory" "files":
 146: resource "azurerm_storage_share_directory" "files" {
```

### Expected Behavior

It should have created a directory structure starting with `Folder1`, then a sub folder `Folder2` and finally `Folder3` underneath that.

```
├── Folder1
│   └── Folder2
│       └── Folder3
```
I would have also expected the following plan.

```hcl
# azurerm_storage_share.files will be created
+ resource "azurerm_storage_share" "files" {
    + id                   = (known after apply)
    + name                 = "files"
    + quota                = 50
    + resource_group_name  = (known after apply)
    + storage_account_name = "jamdevdata01"
    + url                  = (known after apply)
  }

# azurerm_storage_share_directory.files["Folder1"] will be created
+ resource "azurerm_storage_share_directory" "files" {
    + id                   = (known after apply)
    + name                 = "Folder1"
    + share_name           = "files"
    + storage_account_name = "jamdevdata01"
  }

# azurerm_storage_share_directory.files["Folder1/Folder2"] will be created
+ resource "azurerm_storage_share_directory" "files" {
    + id                   = (known after apply)
    + name                 = "Folder1/Folder2"
    + share_name           = "files"
    + storage_account_name = "jamdevdata01"
  }

# azurerm_storage_share_directory.files["Folder1/Folder2/Folder3"] will be created
+ resource "azurerm_storage_share_directory" "files" {
    + id                   = (known after apply)
    + name                 = "Folder1/Folder2/Folder3"
    + share_name           = "files"
    + storage_account_name = "jamdevdata01"
  }
```

### Actual Behavior

It says the directory name has failed validation. However, I believe this is actually caused by this regex not being able to handle a 3 level nested folder structure.

[Storage.go Line 15](https://github.com/terraform-providers/terraform-provider-azurerm/blob/6206160ea000718fa9e564c5829da30949e511d4/azurerm/helpers/validate/storage.go#L15)
```go
if !regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9-]+[A-Za-z0-9]$`).MatchString(value) && !regexp.MustCompile(`^[A-Za-z0-9]{1,}/[A-Za-z0-9]{1,}$`).MatchString(value) {
  errors = append(errors, fmt.Errorf("%s must contain only uppercase and lowercase alphanumeric characters, numbers and hyphens. It must start and end with a letter and end only with a number or letter", k))
}
```

### Steps to Reproduce

1. `terraform init`
2. `terraform apply`

### Important Factoids

I tested this with a 2 level nested structure and Terraform was able to deploy it.

```
├── Folder1
│   └── Folder2
```

### References

* #4165

**Comment by JeremiahInMN** [comment]
This regex mentioned above, also will cause anything with nested folders and a hyphen in the name to fail. 

Example:
This works "folder-name"
This will not work, but should "parent/folder-name"

**Comment by jogleasonjr** [comment]
I added a PR to fix the hyphen bug in the comment from @JeremiahInMN 

I adjusted the regex to allow this provider to accept multiple levels of nested directories, but the Azure API rejected it. This provider would have to implement some looping to get it to work -- not just a regex fix.

```
Error: Error creating Directory "parent/child-two/grand-child" (File Share "fileshare" / Account "redacted"): directories.Client#Create: Failure responding to request: StatusCode=404 -- Original Error: autorest/azure: Service returned an error. Status=404 Code="ParentNotFound" Message="The specified parent path does not exist.\nRequestId:0ea534ae-f01a-00e4-61d8-44c4d8123456\nTime:2020-06-17T18:51:14.1181258Z"
```

**Comment by tombuildsstuff** [comment]
Fixed via #7382
